src/CassetteOverlay.jl

```diff
@@ -120,20 +125,26 @@ macro OverlayPass(method_table::Symbol)
 
     passdef = :(struct $PassName <: $OverlayPass end)
 
-    mtf = (@__MODULE__).method_table
     mtdef = :($CassetteOverlay.method_table(::Type{$PassName}) = $(esc(method_table)))
 
     builtinpass = :(@inline function (::$PassName)(f::Union{Core.Builtin,Core.IntrinsicFunction}, args...)
+        @nospecialize f args
         return f(args...)
     end)
 
     overlaypass = :(@generated function (pass::$PassName)(fargs...)
         return $overlay_generator(pass, fargs)
     end)
 
+    nooverlaytype = typeof(CassetteOverlay.nooverlay)
+    nooverlaydef = :(@inline function (pass::$PassName)(::$nooverlaytype, f, args...)
```

Comment:
why not just `(::OverlayPass)(...)` outside the macro?

Reply:
Do you mean why we want to have the `@OverlayPass` macro?
Or why we want to define `nooverlay` here?

As for the former, it turned out that this generated function based approach requires `CassetteOverlay.method_table(::Type{SomeOverlayPass})` to be defined before we define `@generated (pass::SomeOverlayPass)(...)`.

Reply:
And this `nooverlay` intrinsic is necessary for doing something like:
```julia
# change `sin(x)` to return `cos(x)+sin(x)`
@overlay SinTable sin(x::Union{Float32,Float64}) = cos(x) + nooverlay(sin, x)
```

Reply:
I understand the purpose of the nooverlay intrinsic. What I meant was why does this particular method need to be generated for every subtype of `OverlayPass` rather than just declaring it for the abstract type?

Reply:
This definition will result in method ambiguity error:
```julia
@inline function (::OverlayPass)(::typeof(nooverlay), f, args...)
    @nospecialize f args
    return f(args...)
end
```
```
  MethodError: (::Main.simple.var"##SimpleTable#296")(::typeof(CassetteOverlay.nooverlay), ::typeof(Main.simple.myidentity), ::Nothing) is ambiguous.
  
  Candidates:
    (::CassetteOverlay.OverlayPass)(::typeof(CassetteOverlay.nooverlay), f, args...)
      @ CassetteOverlay ~/julia/packages/CassetteOverlay/src/CassetteOverlay.jl:17
    (var"#34#pass"::Main.simple.var"##SimpleTable#296")(fargs...)
      @ Main.simple ~/julia/packages/CassetteOverlay/src/CassetteOverlay.jl:142
  
  Possible fix, define
    (::Main.simple.var"##SimpleTable#296")(::typeof(CassetteOverlay.nooverlay), ::Any, ::Vararg{Any})
```

Reply:
`CassetteOverlay.method_table(::Type{SomeOverlayPass})` needs to be defined before we define `@generated (pass::SomeOverlayPass)(...)`, otherwise `method_table(passtype)` within `overlay_transform!` will result in world age error (and I guess this is why Cassette's API also uses macro).